src/EventListener/DeserializeListener.php

```diff
@@ -97,7 +97,7 @@ public function onKernelRequest(GetResponseEvent $event): void
         $request->attributes->set(
             'data',
             $this->serializer->deserialize(
-                $requestContent, $context['resource_class'], $format, $context
+                $requestContent, $context['input']['class'] ?? $context['resource_class'], $format, $context + ['api_denormalize' => true]
```

Comment:
Not sure that we should do this here, normally the denormalizer checks if the given class has an input in which case it'll let the denormalization be handled by another normalizer.

Reply:
The basic idea is that we are respecting the contract of the Serializer deserialize. The class we pass into it to deserialize is the class we get out which keeps the AbstractItemNormalizer simpler.

An alternative to this is we update the serializerContextBuilder to add a new parameter like `$context['input_class']` and also set 'api_denormalize` to true so that this calling is a lot easier.

Reply:
> An alternative to this is we update the serializerContextBuilder to add a new parameter like $context['input_class'] and also set 'api_denormalize` to true so that this calling is a lot easier.

Let's see what @teohhanhui thinks about that. I'm afraid that this kind of changes will break some of the current use cases.

Reply:
> The basic idea is that we are respecting the contract of the Serializer deserialize. The class we pass into it to deserialize is the class we get out which keeps the AbstractItemNormalizer simpler.

I like the idea. But yeah, probably it could be done in `SerializerContextBuilder` if it makes things cleaner.

Reply:
But what's the reason of having both `$context['input']['class']` and `$context['input_class']`? Can't we just `unset($context['input'])` instead?

Reply:
@teohhanhui naming could probably be updated, `input_class` is more like `target_class` or `class_to_denormalize`. `['input']['class']` would refer to the actual input class configured for a resource, and `resource_class` would be the actual resource class.

The problem is that the code uses `$context['resource_class']` throughout the codebase.

In the AbstractItemNormalizer, it uses it as either the current class being denormalized or normalized, this is needed because in some of the protected methods that it overrides don't have access to the current class being denormalized/normalized, so using the value from context allows apip to do what it needs to do.

The other way `$context['resource_class']` is as the ACTUAL resource class to get resource metadata from.

Before input/output classes, this was fine because AbstractItemNormalizer only worked on resource classes, so the distinction between those two uses cases wasn't as evident and had no need to change.

But now that we support non-resource classes in AbstractItemNormalizer, we need to keep these fields separate because it breaks the semantics of `resource_class` and how it's used throughout the codebase.